### PR TITLE
[BugFix] fix union rewrite bug

### LINKED ISSUE #31411: union mv rewrite failed
### Steps to reproduce the behavior (Required)
CREATE TABLE `data_type_test_pk` (
 `id` int(11) NOT NULL COMMENT "",
 `create_time` datetime NOT NULL COMMENT "",
 `order_id` bigint(20) NOT NULL COMMENT "",
 `app` varchar(100) NULL COMMENT "",
 `dt` date NOT NULL COMMENT "",
 `amount` decimal64(10, 2) NOT NULL COMMENT "",
 `mill_time` datetime NULL COMMENT "",
 `datetime1` datetime NULL COMMENT "",
 `desc` varchar(500) NULL COMMENT ""
) ENGINE=OLAP 
PRIMARY KEY(`id`, `create_time`)
PARTITION BY RANGE(`create_time`)
(PARTITION p202307 VALUES [("2023-07-01 00:00:00"), ("2023-08-01 00:00:00")),
PARTITION p202308 VALUES [("2023-08-01 00:00:00"), ("2023-09-01 00:00:00")),
PARTITION p202309 VALUES [("2023-09-01 00:00:00"), ("2023-10-01 00:00:00")),
PARTITION p202310 VALUES [("2023-10-01 00:00:00"), ("2023-11-01 00:00:00")))
DISTRIBUTED BY HASH(`id`) BUCKETS 2 
PROPERTIES (
"replication_num" = "1",
"dynamic_partition.enable" = "true",
"dynamic_partition.time_unit" = "month",
"dynamic_partition.time_zone" = "Asia/Shanghai",
"dynamic_partition.start" = "-2147483648",
"dynamic_partition.end" = "1",
"dynamic_partition.prefix" = "p",
"dynamic_partition.buckets" = "2",
"dynamic_partition.history_partition_num" = "0",
"dynamic_partition.start_day_of_month" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT",
"enable_persistent_index" = "true",
"compression" = "LZ4"
); 
​
CREATE TABLE `data_type_test_un` (
 `create_time` datetime NOT NULL COMMENT "",
 `id` bigint(20) NOT NULL COMMENT "",
 `order_id` int(11) NOT NULL COMMENT "",
 `app` varchar(65533) NULL COMMENT "",
 `dt` date NOT NULL COMMENT "业务时间",
 `amount` decimal128(20, 2) NOT NULL COMMENT "",
 `desc` varchar(100) NULL COMMENT "",
 `mill_time` varchar(65533) NULL COMMENT ""
) ENGINE=OLAP 
UNIQUE KEY(`create_time`, `id`)
PARTITION BY RANGE(`create_time`)
(PARTITION p202307 VALUES [("2023-07-01 00:00:00"), ("2023-08-01 00:00:00")),
PARTITION p202308 VALUES [("2023-08-01 00:00:00"), ("2023-09-01 00:00:00")),
PARTITION p202309 VALUES [("2023-09-01 00:00:00"), ("2023-10-01 00:00:00")),
PARTITION p202310 VALUES [("2023-10-01 00:00:00"), ("2023-11-01 00:00:00")))
DISTRIBUTED BY HASH(`id`) BUCKETS 2 
PROPERTIES (
"replication_num" = "1",
"dynamic_partition.enable" = "true",
"dynamic_partition.time_unit" = "month",
"dynamic_partition.time_zone" = "Asia/Shanghai",
"dynamic_partition.start" = "-2",
"dynamic_partition.end" = "1",
"dynamic_partition.prefix" = "p",
"dynamic_partition.buckets" = "2",
"dynamic_partition.history_partition_num" = "0",
"dynamic_partition.start_day_of_month" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT",
"enable_persistent_index" = "false",
"compression" = "LZ4"
); 
​
​
CREATE MATERIALIZED VIEW `view_data_type_test_pk_test2`
PARTITION BY (date_trunc('month', `create_time`))
DISTRIBUTED BY HASH(`id`) BUCKETS 2 
REFRESH ASYNC START("2023-09-12 19:30:00") EVERY(INTERVAL 1 HOUR)
PROPERTIES (
 "replication_num" = "1",
 "storage_medium" = "HDD",
 "partition_ttl_number" = "2"
)
AS SELECT `a`.`id`, `a`.`create_time`, `b`.`create_time` as b_create_time, `a`.`order_id`, `a`.`app`, `a`.`dt`, `a`.`amount`
FROM `data_type_test_pk` AS `a` INNER JOIN `data_type_test_un` AS `b` ON `a`.`id` = `b`.`id`;
​
​
​
INSERT INTO data_type_test_un VALUES ('2023-09-03 09:11:11', '2', '200', '8as', '2023-08-22', '5.20', NULL, NULL);
INSERT INTO data_type_test_un VALUES ('2023-09-03 09:11:11', '1', '200', '8as', '2023-08-22', '5.20', NULL, NULL);
INSERT INTO data_type_test_un VALUES ('2023-07-03 09:11:11', '3', '200', '8as', '2023-08-22', '5.20', NULL, NULL);
INSERT INTO data_type_test_un VALUES ('2023-07-03 09:11:11', '6', '200', '8as', '2023-08-22', '5.20', NULL, NULL);
INSERT INTO data_type_test_un VALUES ('2023-07-03 09:11:11', '7', '200', '8as', '2023-08-22', '5.20', NULL, NULL);
INSERT INTO data_type_test_un VALUES ('2023-07-03 09:11:11', '4', '200', '8as', '2023-08-22', '5.20', NULL, NULL);
​
​
INSERT INTO data_type_test_pk VALUES ('6', '2023-07-31 16:01:45', '600', '6as', '2023-08-22', '5.20', '2023-07-31 16:01:45', '2023-09-03 09:13:13', NULL);
INSERT INTO data_type_test_pk VALUES ('9', '2023-07-31 16:01:45', '600', '6as', '2023-08-22', '5.20', '2023-07-31 16:01:45', '2023-09-03 09:13:13', NULL);
INSERT INTO data_type_test_pk VALUES ('10', '2023-07-31 16:01:45', '600', '6as', '2023-08-22', '5.20', '2023-07-31 16:01:45', '2023-09-03 09:13:13', NULL);
INSERT INTO data_type_test_pk VALUES ('7', '2023-07-31 16:01:45', '600', '6as', '2023-08-22', '5.20', '2023-07-31 16:01:45', '2023-09-03 09:13:13', NULL);
INSERT INTO data_type_test_pk VALUES ('8', '2023-07-31 16:01:45', '600', '6as', '2023-08-22', '5.20', '2023-07-31 16:01:45', '2023-09-03 09:13:13', NULL);
INSERT INTO data_type_test_pk VALUES ('3', '2023-09-03 09:11:11', '300', '3as', '2023-08-21', '3.20', '2023-09-03 09:12:12', '2023-09-03 09:13:13', NULL);
INSERT INTO data_type_test_pk VALUES ('1', '2023-09-03 09:11:11', '100', '1as', '2023-08-19', '1.20', '2023-09-03 09:12:12', '2023-09-03 09:13:13', NULL);
INSERT INTO data_type_test_pk VALUES ('2', '2023-09-03 09:11:11', '200', '2as', '2023-08-20', '2.20', '2023-09-03 09:12:12', '2023-09-03 09:13:13', NULL);
INSERT INTO data_type_test_pk VALUES ('4', '2023-09-03 09:11:11', '400', '4as', '2023-08-22', '4.20', '2023-09-03 09:12:12', '2023-09-03 09:13:13', NULL);
INSERT INTO data_type_test_pk VALUES ('13', '2023-09-03 09:11:11', '400', '4as', '2023-08-22', '4.20', '2023-09-03 09:12:12', '2023-09-03 09:13:13', NULL);
​
​
​
SELECT a.id, a.create_time, a.order_id, a.app, a.dt, a.amount from data_type_test_pk a join data_type_test_un b on a.id = b.id;

### Expected behavior (Required)
query rewritten by mv view_data_type_test_pk_test2
### Real behavior (Required)
query not rewritten by mv view_data_type_test_pk_test2
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

### PR DESCRIPTION
**Pull request by ABingHuang**
there are three reasons:
1. extra compensated partition predicates for none-ref base tables
2. duplicate on predicates should be removed
3. range predicate simplification can be optimized, (a < 10 or a > 100) and a < 50 should be simplied to a < 10

fix it by:
1. remove partition predicates of none-ref base tables
2. normalize on predicate before deduplication
3. enhance range extractor for or predicates

Fixes #31411 

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by murphyatwork — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 100 / 110 (90.91%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|8|11|72.73%|[112, 113, 114]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|6|7|85.71%|[927]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/ColumnRangePredicate.java|60|66|90.91%|[279, 280, 281, 282, 284, 286]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|2|2|100.00%|[]|

File: fe/fe-core/src/test/resources/sql/materialized-view/tpch/q5.sql [hunk 1]
```diff
@@ -29,6 +29,6 @@ TOP-N (order by [[49: sum DESC NULLS LAST]])
         AGGREGATE ([GLOBAL] aggregate [{49: sum=sum(49: sum)}] group by [[42: n_name]] having [null]
             EXCHANGE SHUFFLE[42]
                 AGGREGATE ([LOCAL] aggregate [{49: sum=sum(48: expr)}] group by [[42: n_name]] having [null]
-                    SCAN (mv[lineitem_mv] columns[81: c_nationkey, 96: o_orderdate, 107: s_nationkey, 109: l_saleprice, 115: n_name2, 118: r_name2] predicate[107: s_nationkey = 81: c_nationkey AND 96: o_orderdate >= 1995-01-01 AND 96: o_orderdate < 1996-01-01 AND 118: r_name2 = AFRICA])
+                    SCAN (mv[lineitem_mv] columns[81: c_nationkey, 107: s_nationkey, 109: l_saleprice, 115: n_name2, 118: r_name2] predicate[107: s_nationkey = 81: c_nationkey AND 118: r_name2 = AFRICA])
```
Comment (LiShuMing) [inline fe/fe-core/src/test/resources/sql/materialized-view/tpch/q5.sql:32]: Why remove `o_orderdate >= date '1995-01-01'  and o_orderdate < date '1996-01-01'`?

**Pull request closed without merging**